<a href="https://colab.research.google.com/github/ajengfirdyaramadany-bot/Prediksi-Kualitas-Anggur-One-Quality-Classification-/blob/main/UTS_PDAB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **ULANGAN TENGAH SEMESTER - PENAMBANGAN DATA DAN ANALISIS BISNIS**
Nama : Ajeng Firdya Ramadany
     
NIM : 2404220007

# **1. Persiapan Data**

Pada tahap persiapan data, dilakukan proses pemuatan dan eksplorasi awal terhadap dataset training. Berdasarkan tampilan data, dataset terdiri dari beberapa variabel numerik yang merepresentasikan karakteristik kimia dari wine, seperti fixed acidity, volatile acidity, citric acid, residual sugar, chlorides, free sulfur dioxide, total sulfur dioxide, density, pH, sulphates, dan alcohol.

In [ ]:
import pandas as pd

train = pd.read_csv('/content/data_training.csv')
test = pd.read_csv('/content/data_testing.csv')
template = pd.read_csv('/content/hasilprediksi_3digitNIMterakhir.csv')

In [ ]:
train.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,Id
0,7.3,0.740,0.08,1.7,0.094,10.0,45.0,0.99576,3.24,0.50,9.8,5,1366
1,8.1,0.575,0.22,2.1,0.077,12.0,65.0,0.99670,3.29,0.51,9.2,5,103
2,10.1,0.430,0.40,2.6,0.092,13.0,52.0,0.99834,3.22,0.64,10.0,7,942
3,12.9,0.500,0.55,2.8,0.072,7.0,24.0,1.00012,3.09,0.68,10.9,6,811
4,8.4,0.360,0.32,2.2,0.081,32.0,79.0,0.99640,3.30,0.72,11.0,6,918


Berdasarkan hasil pengamatan, seluruh variabel dalam dataset bersifat numerik, sehingga tidak diperlukan proses encoding data kategorik. Hal ini mempermudah proses analisis dan pemodelan karena data dapat langsung digunakan setelah melalui tahap pembersihan dan preprocessing.

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857 entries, 0 to 856
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         857 non-null    float64
 1   volatile acidity      857 non-null    float64
 2   citric acid           857 non-null    float64
 3   residual sugar        857 non-null    float64
 4   chlorides             857 non-null    float64
 5   free sulfur dioxide   857 non-null    float64
 6   total sulfur dioxide  857 non-null    float64
 7   density               857 non-null    float64
 8   pH                    857 non-null    float64
 9   sulphates             857 non-null    float64
 10  alcohol               857 non-null    float64
 11  quality               857 non-null    int64  
 12  Id                    857 non-null    int64  
dtypes: float64(11), int64(2)
memory usage: 87.2 KB


Meskipun tidak terdapat missing values, variasi skala antar variabel cukup beragam, sehingga diperlukan proses feature scaling pada tahap selanjutnya untuk meningkatkan performa model.

In [ ]:
train.describe()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,Id
count,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000,857.000000
mean,8.261960,0.529393,0.267351,2.506184,0.086830,15.782964,45.978413,0.996692,3.313092,0.656709,10.430338,5.653442,813.749125
std,1.701992,0.179162,0.195144,1.293512,0.048721,10.300402,31.692113,0.001901,0.152079,0.167364,1.066971,0.821777,463.807063
min,4.600000,0.120000,0.000000,0.900000,0.012000,1.000000,6.000000,0.990070,2.740000,0.390000,8.400000,3.000000,0.000000
25%,7.100000,0.395000,0.090000,1.900000,0.070000,7.000000,21.000000,0.995520,3.210000,0.550000,9.500000,5.000000,413.000000
50%,7.900000,0.520000,0.250000,2.200000,0.079000,14.000000,38.000000,0.996680,3.310000,0.620000,10.200000,6.000000,814.000000
75%,9.100000,0.640000,0.420000,2.600000,0.090000,21.000000,63.000000,0.997800,3.400000,0.730000,11.100000,6.000000,1214.000000
max,15.600000,1.580000,1.000000,15.500000,0.611000,68.000000,278.000000,1.003200,4.010000,2.000000,14.000000,8.000000,1597.000000


Berdasarkan hasil statistik deskriptif, dataset terdiri dari 857 observasi untuk setiap variabel, yang menunjukkan bahwa tidak terdapat data yang hilang pada seluruh fitur.

Nilai rata-rata (mean) dari masing-masing variabel menunjukkan karakteristik umum dari data wine. Sebagai contoh, rata-rata nilai fixed acidity sebesar 8.26, alcohol sebesar 10.43, dan pH sebesar 3.31. Sementara itu, variabel target quality memiliki rata-rata sebesar 5.65, yang menunjukkan bahwa sebagian besar kualitas wine berada pada kategori menengah.

Dari sisi penyebaran data, nilai standar deviasi (std) menunjukkan bahwa terdapat variasi pada masing-masing fitur. Variabel seperti total sulfur dioxide memiliki standar deviasi yang cukup besar, yang mengindikasikan adanya penyebaran data yang luas, sedangkan variabel seperti density memiliki standar deviasi kecil, yang berarti nilainya relatif homogen.

Nilai minimum dan maksimum menunjukkan rentang data yang cukup bervariasi. Sebagai contoh, nilai residual sugar berkisar antara 0.9 hingga 15.5, serta alcohol berkisar antara 8.4 hingga 14. Hal ini menunjukkan adanya perbedaan karakteristik yang cukup signifikan antar sampel wine.

# **2. Pembersihan Data**

**A. Pengecekan Missing Values**

Pada tahap ini dilakukan pengecekan terhadap nilai yang hilang pada dataset training dan testing.

In [ ]:
train.isnull().sum()
test.isnull().sum()

,0
fixed acidity,0
volatile acidity,0
citric acid,0
residual sugar,0
chlorides,0
free sulfur dioxide,0
total sulfur dioxide,0
density,0
pH,0
sulphates,0


Interpretasi:
semua nilai 0, artinya tidak terdapat missing values pada seluruh variabel.

**B. Pengecekan Data Duplikat**

Dilakukan pengecekan terhadap data duplikat untuk memastikan tidak ada data yang terduplikasi yang dapat menyebabkan bias dalam pelatihan model.

In [ ]:
train.duplicated().sum()

np.int64(0)

In [ ]:
train = train.drop_duplicates()

Berdasarkan hasil pengecekan data duplikat, diperoleh nilai sebesar 0 yang menunjukkan bahwa tidak terdapat data yang terduplikasi dalam dataset training. Dengan demikian, tidak diperlukan proses penghapusan data duplikat karena seluruh data sudah bersifat unik dan dapat digunakan secara langsung dalam proses pelatihan model.

**C. Pemisahan Fitur dan Target**

Pada tahap ini dilakukan pemisahan antara variabel independen (fitur) dan variabel dependen (target).

In [ ]:
X_train = train.drop(['quality'], axis=1)
y_train = train['quality']

In [ ]:
test_id = test['Id']
X_test = test.drop(['Id'], axis=1)

Variabel quality digunakan sebagai target karena merupakan nilai yang ingin diprediksi dalam permasalahan klasifikasi ini. Sedangkan variabel lainnya digunakan sebagai fitur untuk membangun model.

**D. Feature Scaling**

Feature scaling dilakukan menggunakan metode StandardScaler untuk menstandarisasi nilai fitur sehingga memiliki rata-rata 0 dan standar deviasi 1.

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

# Drop the 'Id' column from X_train to match the features in X_test
# that was created by dropping 'Id'.
# This ensures both X_train and X_test have the same feature set for scaling.
X_train_for_scaling = X_train.drop('Id', axis=1)

X_train_scaled = scaler.fit_transform(X_train_for_scaling)
X_test_scaled = scaler.transform(X_test)

Proses ini penting karena beberapa algoritma machine learning sensitif terhadap skala data. Dengan scaling, model dapat belajar dengan lebih optimal dan konvergen lebih cepat.

# **3. Pembuatan Model**

**a. Import & Training Model**

Model yang digunakan dalam penelitian ini adalah Random Forest Classifier. Model ini dipilih karena mampu menangani data dengan kompleksitas tinggi dan hubungan non-linear antar variabel. Selain itu, Random Forest juga memiliki keunggulan dalam mengurangi overfitting melalui teknik ensemble learning.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(random_state=42)
model.fit(X_train_scaled, y_train)

RandomForestClassifier(random_state=42)

Random Forest Classifier dipilih karena memiliki kemampuan yang baik dalam menangani data numerik dengan pola hubungan yang kompleks dan non-linear. Selain itu, model ini mampu mengurangi risiko overfitting melalui kombinasi banyak decision tree (ensemble learning).

Random Forest juga relatif stabil terhadap noise dan mampu memberikan performa yang baik meskipun terdapat variasi antar fitur pada dataset.

**b. Evaluasi Model**

Pada tahap ini digunakan model Random Forest Classifier untuk melakukan klasifikasi kualitas wine. Model ini dipilih karena mampu menangani hubungan non-linear antar variabel serta memiliki performa yang stabil melalui pendekatan ensemble learning.

Model dilatih menggunakan dataset training yang telah melalui tahap preprocessing, termasuk pembersihan data dan feature scaling. Proses pelatihan bertujuan untuk mempelajari pola hubungan antara variabel fitur dengan variabel target yaitu quality.

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix

y_pred_train = model.predict(X_train_scaled)

print("Accuracy:", accuracy_score(y_train, y_pred_train))
print("Confusion Matrix:\n", confusion_matrix(y_train, y_pred_train))

Accuracy: 1.0
Confusion Matrix:
 [[  6   0   0   0   0   0]
 [  0  26   0   0   0   0]
 [  0   0 362   0   0   0]
 [  0   0   0 341   0   0]
 [  0   0   0   0 109   0]
 [  0   0   0   0   0  13]]


Berdasarkan hasil evaluasi, model menghasilkan nilai akurasi sebesar 1.0 (100%) pada data training. Hal ini menunjukkan bahwa model mampu mengklasifikasikan seluruh data training dengan sangat baik tanpa kesalahan.

Hasil confusion matrix menunjukkan bahwa seluruh nilai berada pada diagonal utama dan tidak terdapat nilai di luar diagonal. Hal ini mengindikasikan bahwa semua data berhasil diprediksi dengan benar oleh model, sehingga tidak terjadi kesalahan klasifikasi antar kelas.

Meskipun demikian, nilai akurasi yang sangat tinggi ini perlu diperhatikan karena berpotensi mengindikasikan overfitting, yaitu kondisi di mana model terlalu menyesuaikan diri dengan data training sehingga kurang mampu melakukan generalisasi pada data baru. Oleh karena itu, diperlukan pengujian lebih lanjut menggunakan data testing untuk memastikan bahwa model tetap memiliki performa yang baik.

# **4. Prediksi Data Testing**

**a. Load Model**

In [ ]:
import joblib

joblib.dump(model, 'model_wine.pkl')

model_loaded = joblib.load('model_wine.pkl')

Model yang telah dilatih kemudian disimpan dalam format .pkl menggunakan library joblib. Proses ini dilakukan agar model dapat digunakan kembali tanpa perlu melakukan pelatihan ulang.

Penyimpanan model merupakan bagian dari deployment sederhana dalam machine learning, karena model dapat diimplementasikan kembali untuk melakukan prediksi pada data baru secara lebih efisien.

**b. Prediksi Data Testing**

In [ ]:
y_pred_test = model_loaded.predict(X_test_scaled)

Setelah model berhasil dimuat kembali, dilakukan proses prediksi terhadap dataset testing menggunakan model Random Forest yang telah dilatih sebelumnya.

Hasil prediksi berupa nilai quality yang merepresentasikan kualitas wine berdasarkan karakteristik kimia yang dimiliki setiap sampel pada data testing.

Proses ini bertujuan untuk mengetahui kemampuan model dalam melakukan klasifikasi pada data baru yang belum pernah digunakan selama proses pelatihan.

In [ ]:
template = pd.read_csv('/content/hasilprediksi_3digitNIMterakhir.csv')

# Isi kolom Quality
template['Quality'] = y_pred_test

**c. Final Data**

In [ ]:
template

,Id;Quality,Quality
0,222;,5
1,1514;,6
2,417;,5
3,754;,5
4,516;,5
...,...,...
281,1147;,6
282,296;,5
283,170;,5
284,1439;,5


Hasil prediksi kemudian disimpan ke dalam file CSV sesuai format yang telah ditentukan, yaitu hanya memuat dua kolom: Id dan Quality.

Kolom Id digunakan sebagai identitas unik setiap data pada dataset testing, sedangkan kolom Quality berisi hasil prediksi kualitas wine yang dihasilkan oleh model Random Forest Classifier.

File ini selanjutnya digunakan sebagai hasil akhir yang akan dibandingkan dengan data kunci untuk mengevaluasi tingkat akurasi prediksi model.

# **Kesimpulan Akhir**

Berdasarkan hasil analisis yang telah dilakukan, model klasifikasi Random Forest Classifier berhasil dibangun untuk memprediksi kualitas wine berdasarkan karakteristik kimia yang terdapat pada dataset.

Proses preprocessing menunjukkan bahwa dataset tidak memiliki missing values maupun data duplikat, sehingga data layak digunakan untuk proses pemodelan. Selain itu, dilakukan feature scaling menggunakan StandardScaler untuk menyeimbangkan skala antar fitur.

Hasil evaluasi model pada data training menunjukkan nilai akurasi sebesar 100% dengan confusion matrix yang seluruh nilainya berada pada diagonal utama. Hal ini menunjukkan bahwa model mampu mempelajari pola data training dengan sangat baik.

Setelah model disimpan dan dimuat kembali menggunakan joblib, model berhasil digunakan untuk memprediksi kualitas wine pada dataset testing. Hasil prediksi kemudian disimpan dalam format CSV sesuai ketentuan yang diberikan.

Secara keseluruhan, model Random Forest Classifier menunjukkan performa yang sangat baik dalam melakukan klasifikasi kualitas wine berdasarkan data yang tersedia.